Detailed guide to deploy Falcon Sensor for Linux on EKS using daemonset (#63)
CrowdStrike/Cloud-AWS: Container Security/kernel-eks-implementation-guide.md

# Implementation Guide for CrowdStrike Falcon Sensor for Linux on EKS Kubernetes cluster using Helm Chart

> This guide works through creation of new Kubernetes cluster, deployment of Falcon Sensor for Linux using Helm Chart, and demonstration of detection capabilities of Falcon Container Workload Protection.

Reviewer: ```suggestion
This guide works through creation of a new Kubernetes cluster, deployment of the Falcon Sensor for Linux using Helm Chart, and demonstration of detection capabilities of Falcon Container Workload Protection.
```

> No prior Kubernetes or Falcon knowledge is needed to follow this guide. First sections of this guide focus on creation of EKS cluster, these sections may be skipped if you have access to an existing cluster.

Reviewer: ```suggestion
No prior Kubernetes or Falcon knowledge is needed to follow this guide. First sections of this guide focus on creation of EKS cluster, however, these sections may be skipped if you have access to an existing cluster.
```

Time needed to follow this guide: 45 minutes.

## Pre-requisites

> You will need AWS credentials and docker tool installed locally.

Reviewer: Which AWS credentials and which docker tools?

## Deployment

### Step 1: Enter the tooling container

 - Install [docker](https://www.docker.com/products/docker-desktop) container runtime
 - Enter the [tooling container](https://github.com/CrowdStrike/cloud-tools-image)
   ```
   docker run --privileged=true \
       -e FALCON_CLIENT_ID="$FALCON_CLIENT_ID" \
       -e FALCON_CLIENT_SECRET="$FALCON_CLIENT_SECRET" \
       -e CID="$CID" \
       -v /var/run/docker.sock:/var/run/docker.sock \
       -v ~/.aws:/root/.aws -it --rm \
       quay.io/crowdstrike/cloud-tools-image

>    ```

Reviewer: Where do the docs tell the reader where/how to set the $FALCON_CLIENT_* vars? Should these be environmental variables or passed natively into the CLI?

>    The above command creates new container runtime that contains tools needed by this guide. All the
>    following commands should be run inside this container. If you have previously used AWS CLI tool,
>    you may already have AWS Credentials stored on your system in `~/.aws` directory. If that is the case,
>    it is preferential to start the container with `-v ~/.aws:/root/.aws:ro` option. This option should be
>    omitted if you don't want to share your AWS credentials with the container. You can review your
>    credentials with `aws sts get-caller-identity` command.
> 
>     Example output
>     ```
>     {
>         "UserId": "AIDAXRCSSEFWMXXXXXXXX",
>         "Account": "123456789123",
>         "Arn": "arn:aws:iam::123456789123:user/xxxxxxx"
>     }
>     ````

Reviewer: ```suggestion
   The above command creates a new container runtime that contains tools needed by this guide. All the
   following commands should be run inside this container. If you have previously used AWS CLI tool,
   you may already have AWS Credentials stored on your system in `~/.aws` directory. If that is the case,
   it is preferential to start the container with `-v ~/.aws:/root/.aws:ro` option to ensure these variables
   are passed through to your container. This option should be omitted if you don't want to share your
   AWS credentials with the container. You can review your credentials with `aws sts get-caller-identity`
   command.

    Example output
    ```
    {
        "UserId": "AIDAXRCSSEFWMXXXXXXXX",
        "Account": "123456789123",
        "Arn": "arn:aws:iam::123456789123:user/xxxxxxx"
    }
    ````
```

### Step 2: Create EKS Cluster

 - Create new EKS cluster. It may take couple minutes before cluster is fully up and functioning.
   ```
   $ eksctl create cluster
       --name demo-cluster --region eu-west-1 \
       --managed
   ```

 - (optional) Verify that your local kubectl utility has been configured to connect to the cluster.
   ```
   $ kubectl cluster-info
   Kubernetes control plane is running at https://EEAB38XXXXXXXXXXXXXXXXXXXXXXXXXX.sk1.eu-west-1.eks.amazonaws.com
   CoreDNS is running at https://EEAB38XXXXXXXXXXXXXXXXXXXXXXXXXX.sk1.eu-west-1.eks.amazonaws.com/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy

   To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
   ```

>    kubectl is command line tool that lets you control Kubernetes clusters. For configuration, kubectl looks for a file named config in the `$HOME/.kube` directory. This config was created previously by `eksctl create cluster` command and contains login information for your newly created cluster.

Reviewer: ```suggestion
   `kubectl` is command line tool that lets you control Kubernetes clusters. For configuration, `kubectl` looks for a file named config in the `$HOME/.kube` directory. This config was created previously by `eksctl create cluster` command and contains login information for your newly created cluster.
```

### Step 3: Create ECR Container Repository

>  - Create container repository in AWS ECR. AWS ECR (Elastic Container Registry) is a cloud service providing a container registry. The bellow command creates a new repository in the registry and this repository will be subsequently used to store the Falcon Node sensor image.

Reviewer: ```suggestion
 - Create container repository in AWS Elastic Container Registry (ECR). AWS ECR is a cloud service providing a container registry. The bellow command creates a new repository in the registry and this repository will be subsequently used to store the Falcon Node sensor image.
```

   ```

>    $ aws ecr create-repository --region eu-west-1 --repository-name falcon-node-sensor
>    {
>        "repository": {
>            "repositoryArn": "arn:aws:ecr:eu-west-1:123456789123:repository/falcon-node-sensor",
>            "registryId": "12345678912345",
>            "repositoryName": "falcon-sensor",
>            "repositoryUri": "123456789123.dkr.ecr.eu-west-1.amazonaws.com/falcon-node-sensor",
>            "createdAt": "2021-02-04T10:30:30+00:00",
>            "imageTagMutability": "MUTABLE",
>            "imageScanningConfiguration": {
>                "scanOnPush": false
>            },
>            "encryptionConfiguration": {
>                "encryptionType": "AES256"
>            }
>        }
>    }
>    ```

Reviewer: ```suggestion
   $ aws ecr create-repository --region eu-west-1 --repository-name falcon-node-sensor
   ```

  This will generate output similar to the following:
  ```
   {
       "repository": {
           "repositoryArn": "arn:aws:ecr:eu-west-1:123456789123:repository/falcon-node-sensor",
           "registryId": "12345678912345",
           "repositoryName": "falcon-sensor",
           "repositoryUri": "123456789123.dkr.ecr.eu-west-1.amazonaws.com/falcon-node-sensor",
           "createdAt": "2021-02-04T10:30:30+00:00",
           "imageTagMutability": "MUTABLE",
           "imageScanningConfiguration": {
               "scanOnPush": false
           },
           "encryptionConfiguration": {
               "encryptionType": "AES256"
           }
       }
   }
   ```
```

Reviewer: Goal here is to clearly show what the user should copy/paste, vs what the output should look like

>  - Note the URI of the newly created repository to the environment variable for further use. Falcon Container Sensor will be available under this URI.

Reviewer: ```suggestion
 - Note the `registryURI` of the newly created repository to the environment variable for further use. Falcon Container Sensor will be available under this URI.
```

   ```
   $ FALCON_NODE_IMAGE_URI=123456789123.dkr.ecr.eu-west-1.amazonaws.com/falcon-node-sensor
   ```

### Step 3: Build falcon-node-sensor container image

>  - Provide OAuth2 Client ID and Client Secret for authentication with CrowdStrike Falcon platform. Establishing and retrieving OAuth2 API credentials can be performed at [falcon-console](https://falcon.crowdstrike.com/support/api-clients-and-keys). These credentials will only be used to download sensor, we recommend you create key pair that has permissions only for

Reviewer: We'll need to complete the sentence around which specific API perms are needed.

   ```
   $ FALCON_CLIENT_ID=1234567890ABCDEFG1234567890ABCDEF
   $ FALCON_CLIENT_SECRET=1234567890ABCDEFG1234567890ABCDEF
   ```

 - (optional) Provide name of Falcon Cloud you want to used
   ```
   $ FALCON_CLOUD=us-1

>    ```

Reviewer: If this is optional, what is the default?

Reviewer: Meaning new customers are /supposed/ to be in us-2. Should us-2 be the default (since that's where customers are heading), or should us-1 be default (since that's where most customers currently are)?

>  - Build falcon-node-sensor container for your particular OS that is running on your cluster nodes. By default EKS clusters use Amazon Linux 2

Reviewer: ```suggestion
 - Build falcon-node-sensor container for your particular OS that is running on your cluster nodes. By default EKS clusters use Amazon Linux 2.
```

   ```
   $ falcon-node-sensor-build amazonlinux2
   ```
   To see various build options see [upstream project](https://github.com/CrowdStrike/Dockerfiles).

### Step 4: Push falcon-node-sensor image to the kube registry

>  - Push the image to the kube registry

Reviewer: ```suggestion
 - Push the image to the kube registry:
```

Reviewer: For consistency, should this be "AWS ECR" or "AWS Container Registry"?

>    ```
>    $ docker tag falcon-node-sensor:latest $FALCON_NODE_IMAGE_URI:latest
>    ```
>    ```
>    $ docker push $FALCON_NODE_IMAGE_URI:latest
>    Using default tag: latest
>    The push refers to repository [123456789123.dkr.ecr.eu-west-1.amazonaws.com/falcon-node-sensor]
>    913513e672ba: Pushed
>    2c5b2be1a19c: Pushed
>    88cd0d79914e: Pushed
>    c20d459170d8: Pushed
>    db978cae6a05: Pushed
>    aeb3f02e9374: Pushed
>    latest: digest: sha256:132eea2728db09c49fecfae78778f11225f6e9818c71c4f5b2321a0dae4d0c95 size: 1572
>    ```

Reviewer: ```suggestion
   ```
   $ docker tag falcon-node-sensor:latest $FALCON_NODE_IMAGE_URI:latest
   ```

  Output will be similar to:
   ```
   $ docker push $FALCON_NODE_IMAGE_URI:latest
   Using default tag: latest
   The push refers to repository [123456789123.dkr.ecr.eu-west-1.amazonaws.com/falcon-node-sensor]
   913513e672ba: Pushed
   2c5b2be1a19c: Pushed
   88cd0d79914e: Pushed
   c20d459170d8: Pushed
   db978cae6a05: Pushed
   aeb3f02e9374: Pushed
   latest: digest: sha256:132eea2728db09c49fecfae78778f11225f6e9818c71c4f5b2321a0dae4d0c95 size: 1572
   ```
```

### Step 5: Deploy the DaemonSet using the helm chart

 - Provide CrowdStrike Falcon Customer ID as environment variable. This CID will be used be helm chart to register your cluster nodes to the CrowdStrike Falcon platform.
   ```
   $ CID=1234567890ABCDEFG1234567890ABCDEF-12
   ```

 - Add the CrowdStrike Falcon Helm repository
   ```
   $ helm repo add crowdstrike https://crowdstrike.github.io/falcon-helm
   ```

>  - Install CrowdStrike Falcon Helm Chart. Above command will install the CrowdStrike Falcon Helm Chart with the release name falcon-helm in the falcon-system namespace.

Reviewer: The test references the above command. What does the command below do? What part is the command, what part is expected output?

   The CrowdStrike Falcon sensor should be spinning up on all your Kubernetes nodes
   now. There should be no further action on your part unless no Falcon Sensor
   container exists in your registry. If you forgot to add a Falcon Sensor image to
   your image registry before you ran `helm install`, please add the Falcon Sensor
   now; otherwise, pods will fail with errors and crash until there is a valid
   image to pull. The default image name to deploy a kernel sensor to a node is
   `falcon-node-sensor`.
   ```
   To learn more about falcon-helm visit [upstream github](https://github.com/CrowdStrike/falcon-helm).

 - (optional) Verify that falcon-node-sensor pod running. You should see one pod per each node on your cluster.
   ```
   $ kubectl get pods -n falcon-system
   NAME                              READY   STATUS    RESTARTS   AGE
   falcon-helm-falcon-sensor-bs98m   2/2     Running   0          21s
   ```
 - (optional) Verify that given pod has registered with CrowdStrike Falcon and received unique identifier.
   ```
   $ kubectl exec -n falcon-system falcon-helm-falcon-sensor-bs98m -c falcon-node-sensor -- falconctl -g --aid
   aid="a582XXXXXXXXXXXXXXXXXXXXXXXXXXXX"

>    ```

Reviewer: could/should we write a one liner that iterates through the `kubectl get pods` command, passes them to this `exec` command, and returns all the AIDs?

Reviewer: Otherwise the use will copy/paste all the kube cluster names. Could get to be a lot of copy/pasting.

 - (optional) Verify that Falcon Sensor for Linux has insert itself to the kernel
    ```
    $ lsmod | grep falcon
    falcon_lsm_serviceable     724992  1
    falcon_nf_netcontain        20480  1
    falcon_kal                  45056  1 falcon_lsm_serviceable
    falcon_lsm_pinned_11110     45056  1
    falcon_lsm_pinned_11308     45056  1
    ```

## Uninstall helm chart

 - Step 1: Uninstall the helm chart
   ```
   helm uninstall falcon-helm -n falcon-system
   ```
 - Step 2: Delete the falcon image from AWS ECR registry
   ```
   $ aws ecr batch-delete-image --region eu-west-1 \
       --repository-name falcon-node-sensor \
       --image-ids imageTag=latest
   {
       "imageIds": [
           {
               "imageDigest": "sha256:e14904d6fd47a8395304cd33a0d650c2b924f1241f0b3561ece8a515c87036df",
               "imageTag": "latest"
           }
       ],
       "failures": []
   }

>    ```

Reviewer: Where do we get the imageDigest value to copy/paste into this command?

Reviewer 2: That looks like the json return value. Like mentioned above it would probably be good to distinguish command vs return values.

 - Step 4: Delete the AWS ECR repository
   ```
   $ aws ecr delete-repository --region eu-west-1 --repository-name falcon-node-sensor
   {
       "repository": {
           "repositoryArn": "arn:aws:ecr:eu-west-1:123456789123:repository/falcon-sensor",
           "registryId": "123456789123",
           "repositoryName": "falcon-sensor",
           "repositoryUri": "123456789123.dkr.ecr.eu-west-1.amazonaws.com/falcon-sensor",
           "createdAt": "2021-02-04T10:30:30+00:00",
           "imageTagMutability": "MUTABLE"
       }
   }

>    ```

Reviewer: Where do we get the repository values to copy/paste into this command?

>  - Step 5: Delete the AWS EKS Fargate Cluster

Reviewer: ```suggestion
 - Step 5: Delete the AWS EKS Fargate Cluster, replacing `eu-west-1` with the AWS region your cluster is running on.
```

Reviewer 2: I think some of the minor corrections Shawn has identified can be committed. Could we merge this? I will walk through the instructions from the main branch here and add/remove on a fork. Trying to get these out to the customer.

Author: Thank You @mccbryan3 and @shawndwells!
